File: demo/static/OxJS/tools/locale/strings.js
var path = Ox.PATH.replace('/build/', '/dev/'),
    files = [
        path + 'Ox/json/Ox.json',
        path + 'Ox.UI/json/Ox.UI.json'
    ],
    strings = {
        'inside Ox._()': {},
        'outside Ox._()': {}
    },
    lines = ['\n'];

Ox.getJSON(files, function(data) {
    files = [];
    Ox.forEach(data, function(data, file) {
        files = files.concat(Ox.flatten(data.files).filter(function(file) {
            return Ox.endsWith(file, '.js');
        }));
    });
    Ox.getAsync(files.map(function(file) {
        return Ox.PATH + file;
    }), Ox.get, function(data) {
        Ox.forEach(data, function(data, file) {
            var inside = false,
                level = 0,
                tokens = Ox.tokenize(data);
            tokens.forEach(function(token, i) {
                var key, string;
                if (i >= 3
                    && tokens[i - 3].value + tokens[i - 2].value
                    + tokens[i - 1].value + tokens[i].value == 'Ox._('
                ) {
                    inside = true;
                    level = 0;
                } else if (inside) {
                    if (token.value == '(') {
                        level++;
                    } else if (token.value == ')') {
                        level--;
                        inside = level > 0;
                    }
                }
                if (token.type == 'string' && token.value.length > 2) {
                    key = inside ? 'inside Ox._()' : 'outside Ox._()';
                    string = token.value.slice(1, -1)
                        .replace(/\\'/g, '\'')
                        .replace(/([^\\]")/, '\\"')
                    if (!Ox.isArray(strings[key][string])) {
                        strings[key][string] = [];
                    }
                    strings[key][string].push(
                        file.replace(path, '') + ':' + token.line
                    );
                }
            });
        });
        Ox.forEach(strings, function(values, key) {
            lines.push(key);
            Ox.sort(Object.keys(values)).forEach(function(string) {
                lines.push(Ox.repeat(' ', 4) + '"' + string + '"');
                Ox.sort(strings[key][string]).forEach(function(file) {
                    lines.push(Ox.repeat(' ', 8) + file);
                });
            });
        });
        Ox.print(lines.join('\n'));
    });
});

File: demo/static/OxJS/tools/locale/strings.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
import os
import ox
import ox.js
import json

strings = {
    'inside Ox._()': {},
    'outside Ox._()': {}
}
lines = []
path = './dev/'
files = []

for f in ['Ox/json/Ox.json', 'Ox.UI/json/Ox.UI.json']:
    with open(path + f) as fd:
        for filename in json.load(fd)['files']:
            if isinstance(filename, list):
                files += filename
            else:
                files.append(filename)

files = [f for f in files if f.endswith('.js')]

for f in files:
    f = os.path.join(path, f)
    with open(f) as fd:
        data = fd.read().decode('utf-8')
        inside = False
        level = 0
        tokens = ox.js.tokenize(data)
        for i, token in enumerate(tokens):
            if i >= 3 and tokens[i - 3]['value'] + tokens[i - 2]['value'] \
                    + tokens[i - 1]['value'] + tokens[i]['value'] == 'Ox._(':
                inside = True
                level = 0
            elif inside:
                if token['value'] == '(':
                    level +=1
                elif token['value'] == ')':
                    level -= 1
                    inside = level > 0
            if token['type'] == 'string' and len(token['value']) > 2:
                key = inside and 'inside Ox._()' or 'outside Ox._()'
                string = token['value'][1:-1].replace(
                        "\\'", "'"
                    ).replace(
                        '\\]"', '\\"'
                    )
                    #.replace(/\\'/g, '\'')
                    #.replace(/([^\\]")/, '\\"')
                if not string in strings[key]:
                    strings[key][string] = []
                strings[key][string].append(
                    f.replace(path, '') + ':%d' % token['line']
                )

for key in ox.sorted_strings(strings):
    lines.append(key)
    for string in ox.sorted_strings(strings[key]):
        lines.append((' ' * 4) + '"%s"' % string)
        for f in ox.sorted_strings(strings[key][string]):
            lines.append((' ' * 8) + f)

print (u'\n'.join(lines)).encode('utf-8')
